Feat/1688 filter dropdown menu

denisonishchenkou opened the pull request:
https://deploy-preview-1409--pattern-lib-unimelb.netlify.com/?selectedKind=Filter%20Dropdown&selectedStory=Default&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel

# Description

FilterDropdown component (work in progress)
NestedCheckbox component (used inside FilterDropdown component, didn't add it to global components)
AppCheckbox component (finished, can be used across apps)

TODO:
Clear/Apply buttons styling

## Type of change

Please delete options that are not relevant.
- [x] New feature (non-breaking change which adds functionality)

# Checklist:

- [ ] Make sure to do not repeat yourself
- [ ] Crossbrowser testing (IE11, Safari 8+, iOS 8.4+, Android 4.4+, Firefox ESR (v52.x), iPhone (4s,6), iPad 2, Galaxy s5) 
- [ ] My changes generate no new warnings
- [ ] I have added tests (e2e and unit) that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Added the new component to the index.js (Vue and Lib) export

denisonishchenkou commented:
@TimMansell could you please help me with checks? Where should I fix the failed semaphoreci build?

And also wanted to ask about the design of the Clear/Apply buttons:

On the desktop view they placed in two columns and on the mobile - in one column, but it looks like they are always 100% width of the column. In this case I can determine when to change the column layout only with ResizeObserver (to get the size of the dropdown), but it doesn't work in Safari (and IE). Therefore, I don't think it's the best solution.
Maybe it's better to change the layout based on viewport resolutions (with media queries)
Or just make them not full width on desktop resolutions?

TimMansell commented:
Hi @denisonishchenkou, a few `.vue` file extensions were missing in the `imports`.  I have fixed them for you.

Fyi, the build was failing on the `yarn build:lib` and `yarn build:vue`

TimMansell commented:
> And also wanted to ask about the design of the Clear/Apply buttons:
> 
> On the desktop view they placed in two columns and on the mobile - in one column, but it looks like they are always 100% width of the column. In this case I can determine when to change the column layout only with ResizeObserver (to get the size of the dropdown), but it doesn't work in Safari (and IE). Therefore, I don't think it's the best solution.
> Maybe it's better to change the layout based on viewport resolutions (with media queries)
> Or just make them not full width on desktop resolutions?

You can use our custom grid for that.  CSS is located in `_grid.css`

html you can use:

```
<div class="grid">
  <div class="cell cell--tab-1of2">button</div>
  <div class="cell cell--tab-1of2">button</div>
</div>
```

TimMansell commented on components/app-checkbox/AppCheckbox.vue:
```diff
@@ -0,0 +1,72 @@
+  <label
+    :class="[
+      'app-checkbox',
+      { 'is-checked': isInputChecked },
```
Instead of adding the class here, can you add it to the element it's targeting and then create a `--modifier` for that class.

ie:

```
<span class="app-checkbox__icon-checked" :class="{ 'app-checkbox__icon-checked--is-checked': isInputChecked}" />
```

TimMansell commented on components/app-checkbox/AppCheckbox.vue:
```diff
@@ -0,0 +1,72 @@
+    :class="[
+      'app-checkbox',
+      { 'is-checked': isInputChecked },
+      { 'is-indeterminate': isIndeterminate },
```
Same with this one.

TimMansell commented on components/app-checkbox/index.css:
```diff
@@ -0,0 +1,84 @@
+    transform: translateX(-100%);
+    opacity: 0;
+
+    .app-checkbox.is-checked & {
```
change to `&--is-checked`

TimMansell commented on components/app-checkbox/index.css:
```diff
@@ -0,0 +1,84 @@
+    transform: translateX(100%);
+    opacity: 0;
+
+    .app-checkbox.is-indeterminate & {
```
change to `&--is-indeterminate`

TimMansell commented on components/app-checkbox/stories/AppCheckboxDefault.vue:
```diff
@@ -0,0 +1,62 @@
+        :label="checkbox.label"
+        :is-checked="typeof checkbox.value === 'boolean' && checkbox.value"
+        :is-indeterminate="checkbox.value === 'indeterminate'"
+        :area-label="checkbox.ariaLabel"
```
Typo.  should be `aria`

TimMansell commented on components/filter-dropdown/FilterDropdown.vue:
```diff
@@ -0,0 +1,134 @@
+<template>
+  <div
+    :class="['filter-dropdown', { 'is-opened': isOpened, 'is-open-up': isOpenUp }]"
```
Instead of adding the classes here add `--modifer` classes to the to the elements that should be targeted.

ie,

- filter-dropdown__select--is-opened
- filter-dropdown__body--is-opened

TimMansell commented on components/filter-dropdown/FilterDropdown.vue:
```diff
@@ -0,0 +1,134 @@
+</template>
+
+<script>
+import _ from 'lodash';
```
Instead of importing the whole lodash libary, import only the required plugins.

```
import cloneDeep from 'lodash.clonedeep';
```

https://www.npmjs.com/package/lodash.clonedeep

TimMansell commented on components/filter-dropdown/components/NestedCheckbox/NestedCheckboxView.vue:
```diff
@@ -0,0 +1,70 @@
+      :key="option.name"
+      :class="[
+        'nested-checkbox',
+        { 'is-first': isFirstNested },
```
convert to a `--modifier`

TimMansell commented on components/filter-dropdown/components/NestedCheckbox/NestedCheckboxUpdateWrapper.vue:
```diff
@@ -0,0 +1,100 @@
+        },
+      };
+
+      _.assign(option, config[newValue]);
```
You can use `Object.assign()`

TimMansell commented on components/filter-dropdown/components/NestedCheckbox/NestedCheckboxUpdateWrapper.vue:
```diff
@@ -0,0 +1,100 @@
+      });
+    },
+    updateParentOptions(options, parentNames) {
+      const parentNamesWithoutLastOption = _.dropRight(parentNames);
```
You can use `.slice()`

TimMansell commented on components/filter-dropdown/nestedCheckboxOptionsValidator.js:
```diff
@@ -0,0 +1,19 @@
+    return false;
+  }
+
+  const validateOptions = (optionsToValidate) => _.every(
```
You can use native js `.every` instead of lodash

TimMansell commented on components/filter-dropdown/components/NestedCheckbox/NestedCheckboxView.vue:
```diff
@@ -0,0 +1,70 @@
+      type: Array,
+      required: true,
+      validator(value) {
+        return _.every(value, (name) => !!name && typeof name === 'string');
```
You can use native js `.every` instead of lodash

TimMansell commented on components/filter-dropdown/components/NestedCheckbox/NestedCheckboxUpdateWrapper.vue:
```diff
@@ -0,0 +1,100 @@
+      _.assign(option, config[newValue]);
+    },
+    updateNestedOptions(options, newValue) {
+      _.each(options, (option) => {
```
you can use `forEach`

gepifanio commented:
Comment: Deploy Started

Building with commit 6a30b230be954ffa3db706dfb825fbb3b7dfe10f

https://app.netlify.com/sites/pattern-lib-unimelb/deploys/5e5fbfb58f5863000867a7ea

gepifanio commented on components/app-checkbox/AppCheckbox.vue:
```diff
@@ -0,0 +1,68 @@
+      :aria-label="areaLabel">
+
+    <span class="app-checkbox__icon">
+      <span :class="['app-checkbox__icon-checked', { 'app-checkbox__icon-checked--is-checked': isInputChecked }]" />
```
Would be better to move to a computed prop

gepifanio replied:
Would be better to move to a computed prop

gepifanio commented on components/app-checkbox/AppCheckbox.vue:
```diff
@@ -0,0 +1,68 @@
+
+    <span class="app-checkbox__icon">
+      <span :class="['app-checkbox__icon-checked', { 'app-checkbox__icon-checked--is-checked': isInputChecked }]" />
+      <span :class="['app-checkbox__icon-indeterminate', { 'app-checkbox__icon-indeterminate--is-indeterminate': isIndeterminate }]" />
```
Would be better to move to a computed prop

gepifanio commented on components/filter-dropdown/FilterDropdown.vue:
```diff
@@ -0,0 +1,167 @@
+    @click.stop>
+    <div
+      ref="select"
+      :class="[
```
would be better to move to a computed prop

gepifanio commented on components/filter-dropdown/FilterDropdown.vue:
```diff
@@ -0,0 +1,167 @@
+
+    <div
+      ref="body"
+      :class="[
```
computed prop

gepifanio commented on components/filter-dropdown/components/NestedCheckbox/NestedCheckboxView.vue:
```diff
@@ -0,0 +1,69 @@
+
+<script>
+import AppCheckbox from '../../../app-checkbox/AppCheckbox.vue';
+// import optionsValidator from './optionsValidator';
```
Remove comment

denisonishchenkou commented:
@gepifanio  I fixed everything that you requested. Could you please take a look, and if nothing else needs to be changed, I will start working on the tests.

gepifanio commented on components/app-checkbox/stories/AppCheckboxDefault.vue:
```diff
@@ -0,0 +1,62 @@
+<template>
```
Could you add in the readme documentation, how to use this component in the Vue component way:

eg:

```
      <app-checkbox
        name="name of the checkbox"
        label="label of the checkbox"...
```

denisonishchenkou replied:
Will do. Also right now AppCheckbox doesn't work with v-model. I can create a wrapper component that will work with it and will be easier to use it across the app (I thought to put current checkbox to the /base folder and use it in a Checkbox that will work with v-model and in a NestedCheckbox). What do you think about that?

denisonishchenkou replied:
Could you please check? I added the readme and tests for the FilterDropdown, NestedCheckbox, AppCheckbox components.

gepifanio replied:
It would be good to make it work with v-model, I guess so would also make it easier to use it once go to the Vue code in the `FAC` project, which basically uses the vue component as well.

gepifanio replied:
The readme looks good, thanks for updating it.

denisonishchenkou commented:
As discussed, I've added AppCheckbox (that works with `v-model`) and updated BaseCheckbox component.
Also added readme and tests.

@gepifanio could you please review?

gepifanio commented:
> As discussed, I've added AppCheckbox (that works with `v-model`) and updated BaseCheckbox component.
> Also added readme and tests.
> 
> @gepifanio could you please review?

He @denisonishchenkou, looking good so far for me, well done, doing a good job!

denisonishchenkou commented:
@gepifanio do I need to do anything else in this PR before merge? If everything is ok, I can start working in the `find-a-course` repo after this PR is merged

TimMansell commented:
> @gepifanio do I need to do anything else in this PR before merge? If everything is ok, I can start working in the `find-a-course` repo after this PR is merged

Hi @denisonishchenkou, I have approved and merged this PR.  Can you work on combining all the components into one story on #1414 so that:

- It can be demonstrated all working together
- Can easily be tested by us and approved for release

![Screen Shot 2020-03-26 at 9 39 00 am](https://user-images.githubusercontent.com/1926461/77592142-a4691c80-6f45-11ea-8ecd-b9978c56089d.png)

Don't worry about the design of the cards (below) as that will be implemented in `find-a-course`.  You can use the same cards that is in https://deploy-preview-1414--pattern-lib-unimelb.netlify.com/?selectedKind=Filtered%20Results&selectedStory=With%20results&full=0&addons=1&stories=1&panelRight=0&addonPanel=REACT_STORYBOOK%2Freadme%2Fpanel

![Screen Shot 2020-03-26 at 9 37 06 am](https://user-images.githubusercontent.com/1926461/77592025-5e13bd80-6f45-11ea-999e-c157185952ac.png)
